exonum/src/runtime/rust/api.rs

```diff
@@ -160,10 +162,120 @@ impl ServiceApiScope {
+                },
+            );
+            let result = handler(&state, query);
+            Box::new(result.into_future())
+        };
+        // Mark endpoint as deprecated.
+        let mut handler = Deprecated::from(handler);
+        if let Some(deprecates_on) = deprecates_on {
+            handler = handler.with_date(deprecates_on);
+        }
+        if let Some(description) = description {
+            handler = handler.with_description(description);
+        }
+        self.inner.endpoint(name, handler);
+        self
+    }
+
+    /// Same as `endpoint_mut`, but the response will contain a warning about endpoint being deprecated.
+    /// Optional endpoint expiration date and deprecation-related information (e.g., link to a documentation for
+    /// a new API) can be included in the warning.
+    pub fn deprecated_endpoint_mut<Q, I, F, R>(
+        &mut self,
+        name: &'static str,
+        deprecates_on: Option<DateTime<Utc>>,
+        description: Option<String>,
+        handler: F,
+    ) -> &mut Self
+    where
+        Q: DeserializeOwned + 'static,
+        I: Serialize + 'static,
+        F: Fn(&ServiceApiState<'_>, Q) -> R + 'static + Clone + Send + Sync,
+        R: IntoFuture<Item = I, Error = crate::api::Error> + 'static,
+    {
+        let blockchain = self.blockchain.clone();
+        let descriptor = self.descriptor.clone();
+        let handler = move |query: Q| -> crate::api::FutureResult<I> {
+            let state = ServiceApiState::from_api_context(
+                &blockchain,
+                InstanceDescriptor {
+                    id: descriptor.0,
+                    name: descriptor.1.as_ref(),
+                },
+            );
+            let result = handler(&state, query);
+            Box::new(result.into_future())
+        };
+        // Mark endpoint as deprecated.
+        let mut handler = Deprecated::from(handler);
+        if let Some(deprecates_on) = deprecates_on {
+            handler = handler.with_date(deprecates_on);
+        }
+        if let Some(description) = description {
+            handler = handler.with_description(description);
+        }
+        self.inner.endpoint_mut(name, handler);
+        self
+    }
+
     /// Return a mutable reference to the underlying web backend.
     pub fn web_backend(&mut self) -> &mut crate::api::backends::actix::ApiBuilder {
         self.inner.web_backend()
     }
+
+    /// Takes an old endpoint and a new endpoint as direct URIs, and creates
+    /// a relative path from the old to the new one.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::runtime::rust::api::ServiceApiScope;
+    ///
+    /// fn main() {
+    ///    let relative_path = ServiceApiScope::relative_to("hello/world", "exonum/rust/app");
+    ///    assert_eq!(&relative_path, "../../exonum/rust/app");
+    /// }
+    /// ```
+    pub fn relative_to(old_endpoint: &str, new_endpoint: &str) -> String {
```

Comment:
What if the endpoint path (relative to the service root) is a part of `ServiceApiState`? In this case, it would possible to encapsulate this logic in something like

```rust
scope.endpoint("v1/foo", |state, query| {
    let new_query = // based on `query`
    Err(state.moved_permanently("v2/foo").with_query(&new_query))
})
```

Reply:
Done, now it's possible to create `MovedPermanently` response with builder.